File: 169-Majority Element.py
# 排序
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums.sort()
        return nums[int(len(nums) / 2)]


'''
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        checked, len_ = [], len(nums)
        for num in nums:
            if num not in checked and nums.count(num) > len_ / 2: return num
            checked.append(num)
'''

File: C++/169-Majority_Element.cpp
// Runtime: 28 ms, faster than 26.89% of C++ online submissions for Majority Element.
// Memory Usage: 11 MB, less than 93.94% of C++ online submissions for Majority Element.
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        return nums[nums.size() / 2];
    }
};


// Runtime: 16 ms, faster than 96.74% of C++ online submissions for Majority Element.
// Memory Usage: 11 MB, less than 96.97% of C++ online submissions for Majority Element.
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int res = nums[0], times = 1;
        for (int i = 1; i < nums.size(); ++i) {
            if (res == nums[i])
                ++times;
            else if (times > 0)
                --times;
            else
                res = nums[i];
        }
        return res;
    }
};
